src/components/tl/nccl/allgatherv/allgatherv.c

```diff
@@ -0,0 +1,270 @@
+    size_t              max_count, sdt_size, rdt_size;
+    ucc_rank_t          peer;
+
+    CHECK_INPLACE(*args, nccl_team);
+    CHECK_USERDEFINED_DT(*args, nccl_team);
+    task = ucc_tl_nccl_init_task(coll_args, team);
+    if (ucc_unlikely(!task)) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    sdt_size = ucc_dt_size(args->src.info.datatype);
+    rdt_size = ucc_dt_size(args->dst.info_v.datatype);
+    max_count = ucc_coll_args_get_count(args, args->dst.info_v.counts, 0);
+    for (peer = 1; peer < team->params.size; peer++) {
+        max_count = ucc_max(ucc_coll_args_get_count(args,
+                            args->dst.info_v.counts, 0), max_count);
+    }
+    if (max_count * rdt_size > args->src.info.count * sdt_size) {
+        status = ucc_mc_alloc(&task->scratch, (team->params.size + 1) *
+                              max_count *
+                              ucc_dt_size(args->dst.info_v.datatype),
+                              UCC_MEMORY_TYPE_CUDA);
+
+    } else {
+        status = ucc_mc_alloc(&task->scratch, max_count * team->params.size *
+                              ucc_dt_size(args->dst.info_v.datatype),
+                              UCC_MEMORY_TYPE_CUDA);
+    }
+    if (ucc_unlikely(status != UCC_OK)) {
+        ucc_tl_nccl_free_task(task);
+        return status;
+    }
+    task->super.post = ucc_tl_nccl_allgatherv_bcopy_start;
+    *task_h = &task->super;
+out:
+    return status;
+}
+
+ucc_status_t ucc_tl_nccl_allgatherv_bcast_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task   = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_coll_args_t    *args   = &TASK_ARGS(task);
+    ucc_tl_nccl_team_t *team   = TASK_TEAM(task);
+    ucc_rank_t          size   = UCC_TL_TEAM_SIZE(team);
+    ucc_ee_h            ee     = coll_task->ee;
+    cudaStream_t        stream = (ee) ? (cudaStream_t) ee->ee_context :
+                                                       team->stream;
+    ucc_status_t        status = UCC_OK;
+    void               *sbuf   = args->src.info.buffer;
+    ptrdiff_t           rbuf   = (ptrdiff_t)args->dst.info_v.buffer;
+    size_t rdt_size, count, displ;
+    ucc_rank_t peer;
+
+    task->super.super.status = UCC_INPROGRESS;
+    rdt_size                 = ucc_dt_size(args->dst.info_v.datatype);
+    UCC_TL_NCCL_PROFILE_REQUEST_EVENT(coll_task, "nccl_allgatherv_start", 0);
+    NCCLCHECK_GOTO(ncclGroupStart(), exit_coll, status, UCC_TL_TEAM_LIB(team));
+    for (peer = 0; peer < size; peer++) {
+        count = ucc_coll_args_get_count(args, args->dst.info_v.counts, peer);
+        displ = ucc_coll_args_get_displacement(args,
+                                               args->dst.info_v.displacements,
+                                               peer);
+        NCCLCHECK_GOTO(ncclBroadcast(sbuf, PTR_OFFSET(rbuf, displ * rdt_size),
+                                     count * rdt_size, ncclChar, peer,
+                                     team->nccl_comm, stream),
```

Comment:
is this correct given that scount is different from rcount for a peer?

Reply:
i think it's correct, rcount[R] * rdt_size should be equal to count * sdt_size for rank R, or I'm missing something?

Reply:
in allgatherv isn't each process send the same input to all ranks and receives different lengths from different ranks?
rank 0 can send XX bytes and receive YY bytes from rank PEER.